realm/src/main/java/io/realm/Realm.java

```diff
@@ -287,9 +313,235 @@ private void removeHandler(Handler handler) {
     private class RealmCallback implements Handler.Callback {
         @Override
         public boolean handleMessage(Message message) {
-            if (message.what == REALM_CHANGED) {
-                transaction.advanceRead();
-                sendNotifications();
+            switch (message.what) {
+                case REALM_CHANGED: {
+                    if (threadContainsAsyncQueries()) {
```

Comment:
Also I would probably also refactor `if/else` branches that span that many line into separate methods so it is easier to make out the control flow.